### PR TITLE
[PDPD FE] : Refactored  PaddlePaddle Quantization

### PR DESCRIPTION
**Pull request by Pey-crypto**
Fix for #20687
1. Supports both per-tensor and per-channel quantization using the axis attribute.
2. Uses FakeQuantize for quantize_linear and a simple subtract-multiply operation for dequantize_linear.
3. The HALF_AWAY_FROM_ZERO round mode is ignored, as it's not explicitly handled.

### CODE REVIEW

**Comment by xczhai** [comment]
Hi @Pey-crypto 
Thanks for your commits. But we can polish this PR for better maintain.
1. please check https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/paddle/src/internal/pass/transform_fakequantize.cpp is needed or not. Theoretically, such pass should be remove since your commit works.
2. Previously, I prepare some unittests for Quantization. Maybe you can import these tests to such PR. It can ensure the motification accuracy. https://github.com/xczhai/openvino/tree/refactor_paddle_slim
BR.

**Comment by Pey-crypto** [comment]
@xczhai ,
Currently working on the errors.

**Comment by xczhai** [comment]
> @xczhai , Currently working on the errors.

If you have any questions, feel free reach me. I will reply asap.

**Comment by xczhai** [comment]
> > > @xczhai , Currently working on the errors.
> > 
> > 
> > If you have any questions, feel free reach me. I will reply asap.
> 
> https://github.com/xczhai/openvino/blob/a839d4c87e6275851f8e08d518d1e0049d382adb/src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py Is this the correct file?. I thought it would be a c++ test file rather than python.

For 

> > > @xczhai , Currently working on the errors.
> > 
> > 
> > If you have any questions, feel free reach me. I will reply asap.
> 
> https://github.com/xczhai/openvino/blob/a839d4c87e6275851f8e08d518d1e0049d382adb/src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py Is this the correct file?. I thought it would be a c++ test file rather than python.



> > > @xczhai , Currently working on the errors.
> > 
> > 
> > If you have any questions, feel free reach me. I will reply asap.
> 
> https://github.com/xczhai/openvino/blob/a839d4c87e6275851f8e08d518d1e0049d382adb/src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py Is this the correct file?. I thought it would be a c++ test file rather than python.

Hi, 
- The test case and paddle api may be out of date. I have attached the reference link in the tests. If the api is lagacy, you can refer to those paddle repo links. https://github.com/xczhai/openvino/blob/a839d4c87e6275851f8e08d518d1e0049d382adb/src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py#L27
- it is acceptable to cover the OP mapping correctness under some paddle model snippet for testing. C++ test is not needed here. You can refer to https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/paddle/tests/test_models/gen_scripts.

**Comment by Pey-crypto** [comment]
@xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help.

**Comment by xczhai** [comment]
> @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.

Did you execute "it submodule update --init --recursive" before building codes?

**Comment by Pey-crypto** [comment]
> > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> 
> Did you execute `git submodule update --init --recursive` before building codes?

yup multiple times, i even cleared the submodules again and tried it again.

**Comment by xczhai** [comment]
> > > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> > 
> > 
> > Did you execute `git submodule update --init --recursive` before building codes?
> 
> yup multiple times, i even cleared the submodules again and tried it again. I even used the --progress and --verbose flags to see the git clone process underneath, but all the clones were happening successfully

please share the commands you have tried.

**Comment by Pey-crypto** [comment]
> > > > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> > > 
> > > 
> > > Did you execute `git submodule update --init --recursive` before building codes?
> > 
> > 
> > yup multiple times, i even cleared the submodules again and tried it again. I even used the --progress and --verbose flags to see the git clone process underneath, but all the clones were happening successfully
> 
> please share the commands you have tried.

git submodule update --init --recursive --progress

**Comment by Pey-crypto** [comment]
> > > > > > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> > > > > 
> > > > > 
> > > > > Did you execute `git submodule update --init --recursive` before building codes?
> > > > 
> > > > 
> > > > yup multiple times, i even cleared the submodules again and tried it again. I even used the --progress and --verbose flags to see the git clone process underneath, but all the clones were happening successfully
> > > 
> > > 
> > > please share the commands you have tried.
> > 
> > 
> > git submodule update --init --recursive --progress cmake --build . --parallel $(nproc) cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_MLAS=OFF -DENABLE_ONEDNN=OFF ..
> 
> You should turn on mlas and onednn though it takes more time to build.

i did also, but then cmake reports mlas and onednn makefiles aren't available. I will try again and get back with the results.

**Comment by xczhai** [comment]
> > > > > > > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> > > > > > 
> > > > > > 
> > > > > > Did you execute `git submodule update --init --recursive` before building codes?
> > > > > 
> > > > > 
> > > > > yup multiple times, i even cleared the submodules again and tried it again. I even used the --progress and --verbose flags to see the git clone process underneath, but all the clones were happening successfully
> > > > 
> > > > 
> > > > please share the commands you have tried.
> > > 
> > > 
> > > git submodule update --init --recursive --progress cmake --build . --parallel $(nproc) cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_MLAS=OFF -DENABLE_ONEDNN=OFF ..
> > 
> > 
> > You should turn on mlas and onednn though it takes more time to build.
> 
> i did do that also, but then cmake reports mlas and onednn makefiles aren't available. I will try again and get back with the results.

Could you share the CPU platform? I can try to reproduce it locally.

**Comment by Pey-crypto** [comment]
> > > > > > > > @xczhai I am not able to build or test the commit I have made , because mlas and oneDNN are missing their MAKE files during the build. Using the -DENABLE flags doesn't seem to help. I have made two unit tests, for the functions and have not changed the underlying python api.
> > > > > > > 
> > > > > > > 
> > > > > > > Did you execute `git submodule update --init --recursive` before building codes?
> > > > > > 
> > > > > > 
> > > > > > yup multiple times, i even cleared the submodules again and tried it again. I even used the --progress and --verbose flags to see the git clone process underneath, but all the clones were happening successfully
> > > > > 
> > > > > 
> > > > > please share the commands you have tried.
> > > > 
> > > > 
> > > > git submodule update --init --recursive --progress cmake --build . --parallel $(nproc) cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_MLAS=OFF -DENABLE_ONEDNN=OFF ..
> > > 
> > > 
> > > You should turn on mlas and onednn though it takes more time to build.
> > 
> > 
> > i did do that also, but then cmake reports mlas and onednn makefiles aren't available. I will try again and get back with the results.
> 
> Could you share the CPU platform and OS info? I can try to reproduce it locally.

Operating System: Ubuntu 24.04 LTS
Model name:            12th Gen Intel(R) Core(TM) i5-12500H
Architecture: x86-64

**Comment by Pey-crypto** [comment]
@xczhai Finally figured out the error, the permissions on those folders were messed up somehow, attaching strace revealed perms were being changed. I guess it was local issue on my end.

**Comment by Pey-crypto** [comment]
> Hi @Pey-crypto Thanks for your commits. But we can polish this PR for better maintain.
> 
>     1. please check https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/paddle/src/internal/pass/transform_fakequantize.cpp is needed or not. Theoretically, such pass should be remove since your commit works.
> 
>     2. Previously, I prepare some unittests for Quantization. Maybe you can import these tests to such PR. It can ensure the motification accuracy. https://github.com/xczhai/openvino/tree/refactor_paddle_slim
> 
> 
> BR.

Removed transform_fakquantize as the node is being made in quantize_linear. Removed references from the paddle frontend.

**Comment by mlukasze** [comment]
@xczhai please evaluate current state of the PR, enable jenkins pipelines if good enough

**Comment by Pey-crypto** [comment]
@xczhai I used structured bindings, which wasn't compliant with C++ 11 standard, I have replaced it with tie.

**Comment by xczhai** [comment]
> @xczhai I used structured bindings, which wasn't compliant with C++ 11 standard, I have replaced it with tie.

Yes, you are right. The current implementation should be aligned with C++11.

**Comment by Pey-crypto** [comment]
@xczhai , I have fixed the tests. Could you review it?

**Comment by xczhai** [comment]
> @xczhai , I have fixed the tests. Could you review it?

OK. I start it.

File: src/frontends/paddle/CMakeLists.txt [hunk 1]
```diff
@@ -2,8 +2,9 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
-add_subdirectory(src)
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../../core/dev_api)
 
+add_subdirectory(src)
```
Comment (xczhai) [inline src/frontends/paddle/CMakeLists.txt:7]: why need to modify this CMakeList?

Comment (Pey-crypto) [reply src/frontends/paddle/CMakeLists.txt:7]: I wanted to access a utility func in the validation_util header file in the core_dev folder. I modified the file to tell MakeTool the location of the file.

Comment (xczhai) [reply src/frontends/paddle/CMakeLists.txt:7]: @Pey-crypto 
I see. Do you want to call `ov::util::normailize_axis` in the `openvino/core/validation_util.hpp`? If so, `ov::util::normailize_axis` is a very simple function, which is used to correct axis when it is negative. You can implement it like `(value < 0) ? value + max : value;`.
Need not to modify the Makefile and import more dependency.

File: src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
+import numpy as np
+import paddle
+from save_model import exportModel
+import sys
+
```
Comment (xczhai) [inline src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py:5]:
```suggestion
from paddle.base.framework import in_dygraph_mode
if paddle.__version__ >= '2.6.0':
    from paddle.base.layer_helper import LayerHelper
else:
    from paddle.fluid.layer_helper import LayerHelper
```

File: src/frontends/paddle/tests/test_models/gen_scripts/generate_dequantize_linear.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
+import numpy as np
+import paddle
+from save_model import exportModel
+import sys
+
```
Comment (xczhai) [inline src/frontends/paddle/tests/test_models/gen_scripts/generate_dequantize_linear.py:5]:
```suggestion
from paddle.base.framework import in_dygraph_mode
if paddle.__version__ >= '2.6.0':
    from paddle.base.layer_helper import LayerHelper
else:
    from paddle.fluid.layer_helper import LayerHelper
```

File: src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    # Test: Per-channel quantization
+    quant_axis = 1
+    scale = [0.1, 0.2, 0.3]
+    zero_point = [0, 0, 0]
+    test_quantize_linear("quantize_linear_per_channel", input_data, quant_axis, scale=scale, zero_point=zero_point)
+
+    # Test: Different bit lengths
+    quant_axis = -1
+    test_quantize_linear("quantize_linear_4bit", input_data, quant_axis, bit_length=4)
```
Comment (xczhai) [inline src/frontends/paddle/tests/test_models/gen_scripts/generate_quantize_linear.py:92]: For now, we don't support 4-bit. can remove this test.

**Comment by xczhai** [comment]
@Pey-crypto 
Two things need to do.

1. need to fix the testcase. for example.
```bash
cd //openvino/src/frontends/paddle/tests
python test_models/gen_scripts/generate_ceil.py ///openvino/bin/intel64/Debug/test_model_zoo/paddle_test_models/

cd //openvino/bin/intel64/Debug
./paddle_tests --gtest_filter=PaddleFuzzyOpTest/FrontEndFuzzyOpTest.testOpFuzzy/paddle_ceil_*
```
2. need to check the executable graph and compare the performance.
```bash
cd //openvino/bin/intel64/Debug
./benchmark -m xxx.pdmodel # can measure the perf
./benchmark -m xxx.pdmodel -exec_graph_path "exec_graph.xml"  # can dump the fused graph, which can be open in netron
```

**Comment by mlukasze** [comment]
@Pey-crypto will you have a time to address issues highlighted in recent comment?

**Comment by Pey-crypto** [comment]
> @Pey-crypto will you have a time to address issues highlighted in recent comment?

Yes will do within this week, there were festivities😁

**Comment by mlukasze** [comment]
that's awesome, looking forward for your commits then!

File: src/frontends/onnx/frontend/src/op/dequantize_linear.cpp
Comment (xczhai) [inline src/frontends/onnx/frontend/src/op/dequantize_linear.cpp:1]: Should not modify onnx implementation. ONNX and PDPD are seperate frontend.

**Pull request closed without merging**